beetmoverscript/src/beetmoverscript/script.py

```diff
@@ -285,6 +290,208 @@ async def upload_data(context):
+             "destinations": [
+               "logdest",
+             ]
+           },
+           "*": {
+             "destinations": [
+               "dest1",
+             ]
+           }
+         },
+       }
+     ]
+
+     The return value takes the form of:
+     [
+         {
+             "taskId": "taskId1",
+             "paths": {
+                 "/path/to/public/build/file": {
+                     "destinations": [
+                         "dest1/file",
+                     ]
+                 },
+                 "/path/to/public/build/file2": {
+                     "destinations": [
+                         "dest1/file2",
+                     ]
+                 },
+                 "/path/to/public/logs/live.log": {
+                     "destinations": [
+                         "dest1/live.log",
+                     ]
+                 },
+             }
+         },
+         {
+             "taskId": "taskId2",
+             "paths": {
+                 "/path/to/public/build/file3": {
+                     "destinations": [
+                         "dest1/file3",
+                     ]
+                 },
+                 "/path/to/public/logs/live.log": {
+                     "destinations": [
+                         "logdest/live.log",
+                     ]
+                 },
+             }
+         }
+     ]
+
+     Each artifact may match up to two `paths` in the `artifactMap`: a `*` path
+     and one other path. If an artifact matches both of these, the non-`*` one
+     takes precedence, and it will _not_ be written to the `*` path destinations.
+     This allows us to have specific filenames (or extensions, prefixes, etc.)
+     grouped together, and to have a "catch-all" for any unmatched artifacts.
+
+     Sanity checking is done to ensure that no destination files would overwrite
+     each other.
+    """
+
+    concreteArtifactMap = []
+    errors = []
+
+    for map_ in artifactMap:
+        concretePaths = {}
+
+        full_glob_destinations = map_["paths"].get("*", {}).get("destinations")
+        other_paths = copy.deepcopy(map_["paths"])
+        if "*" in other_paths:
+            del other_paths["*"]
+
+        for taskId, artifacts in upstreamArtifactPaths.items():
+            if map_["taskId"] != taskId:
+                continue
+
+            for artifact in artifacts:
+                dest_artifact = remove_segments(artifact, strip_segments)
+
+                destinations = []
+                # We need to look at non-'*' paths separate from '*' paths.
+
+                for input_path, output in other_paths.items():
+                    # Skip any input paths that don't match the artifact name.
+                    if "*" in input_path:
+                        if not fnmatch.fnmatch(dest_artifact, input_path):
+                            continue
+                    else:
+                        if input_path != artifact:
+                            continue
```

Comment:
Is there a reason we need to treat globs and non-globs differently here, where one is compared with `artifact` and the other with `dest_artifact`?  Couldn't we just always use fnmatch against either (not sure which makes more sense)

Reply:
We probably could come to think of it; I'll give it a try.